File: Algorithms/基础/栈/单调栈/LC907.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"
#define lowbit(x) (x & (-x))
typedef long long LL;
typedef unsigned long long ULL;
typedef pair<int, int> PII;
const int INF = 0x3f3f3f3f, MOD = 1e9 + 7;

// https://leetcode.cn/problems/sum-of-subarray-minimums/

class Solution {
public:
    typedef long long LL;
    const int MOD = 1e9 + 7;
    int sumSubarrayMins(vector<int>& arr) {
        int n = arr.size();
        vector<int> L(n, -1), R(n, n);
        stack<int> st;
        for (int i = 0; i < n; ++i) {
            while (st.size() && arr[i] < arr[st.top()]) {
                R[st.top()] = i;
                st.pop();
            }
            if (st.size()) L[i] = st.top();
            st.push(i);
        }
        int res = 0;
        for (int i = 0; i < n; ++i)
            res = (res + (LL)(i - L[i]) * (R[i] - i) * arr[i]) % MOD;
        return res;
    }
};

int main() {
    return 0;
}

File: Algorithms/基础/栈/单调栈/LC907.go
package main

// https://leetcode.cn/problems/sum-of-subarray-minimums/?envType=daily-question&envId=2023-11-27

func sumSubarrayMins(arr []int) (res int) {
	const MOD int = 1e9 + 7
	n := len(arr)
	L, R := make([]int, n), make([]int, n)
	for i := 0; i < n; i++ {
		L[i], R[i] = -1, n
	}
	st := []int{}
	for i, v := range arr {
		for len(st) > 0 && arr[st[len(st)-1]] > v {
			R[st[len(st)-1]] = i
			st = st[:len(st)-1]
		}
		if len(st) > 0 {
			L[i] = st[len(st)-1]
		}
		st = append(st, i)
	}
	for i := 0; i < n; i++ {
		res += (i - L[i]) * (R[i] - i) * arr[i]
		res %= MOD
	}
	return
}
